yt_idefix/data_structures.py

```diff
@@ -320,3 +321,98 @@ def _parse_parameter_file(self):
+        units_override=None,
+        *,
+        code_dir: str | None = None,
+        geometry: str | None = None,
+        inifile=None,
+    ):
+        # Presuming code files are along with output, otherwise header file location
+        # has to be specified.
+        if code_dir is None:
+            self.code_dir = os.path.abspath(os.path.dirname(filename))
+        else:
+            self.code_dir = os.path.abspath(code_dir)
+
+        super().__init__(
+            filename,
+            dataset_type=dataset_type,
+            unit_system=unit_system,
+            units_override=units_override,
+            geometry=self.geometry,
+            inifile=inifile,
+        )
+
+    def _parse_parameter_file(self):
+        self._parse_header_file()
+        super()._parse_parameter_file()
+        self._get_time()
+
+    def _parse_header_file(self):
+        """Pluto reads some metadata from header file 'definitions.h'."""
+        constexpr = re.compile(r"CONST_\w*")
+        header_file = os.path.join(self.code_dir, "definitions.h")
+        if os.path.isfile(header_file):
+            with open(header_file) as hf:
+                for line in hf.readlines():
+                    if "GEOMETRY" in line:
+                        self.geometry = line.split()[-1].lower()
+                    if "UNIT" in line:
+                        item = line.split()
+                        key = item[1].lower()
+                        expr = item[-1]
+                        if "CONST" in expr:
+                            expr = constexpr.sub(self._get_const_for_reg, expr)
+                        self.parameters[key] = eval(expr)
+        elif self.geometry is None:
+            raise ValueError(
+                f"Header file couldn't be found in {self.code_dir}. "
+                "The 'geometry' keyword argument must be specified. "
+                "The code units are set to be 1.0 in cgs by default."
+            )
+        else:
+            warnings.warn(
+                f"Header file couldn't be found in {self.code_dir}. "
+                "The code units are set to be 1.0 in cgs by default."
+            )
+
+    def _get_const_for_reg(self, matched):
+        """Replace matched constant string with value"""
+        key = matched.group()
+        return str(pluto_def_constants[key])
+
+    def _get_time(self):
+        """Get current time from vtk.out."""
+        log_file = os.path.join(self.directory, "vtk.out")
+        index = int(re.search(r"\.(\d*)\.", self.parameter_filename).group(1))
+        if os.path.isfile(log_file):
+            with open(log_file) as log:
+                for line in log.readlines():
+                    if line.startswith(str(index)):
+                        self.current_time = np.float32(line.split()[1])
+        else:
+            warnings.warn(
+                f"Log file couldn't be found in {self.fullpath}. "
+                "The 'current_time' attributte is set to be -1."
+            )
```

Comment:
this doesn't mandate as a `warnings.warn` call in my opinion because it's unlikely that users getting the warning will be able to act on it.
I'd go with silently setting the default value (which seems to be missing here actually)

Reply:
> it's unlikely that users getting the warning will be able to act on it.

Although the value of `current_time` is printed by yt_logger after `load`, I thought it can tell users the `current_time` isn't set correctly due to lacking log file, not due to other bugs.

> I'd go with silently setting the default value (which seems to be missing here actually)

Because `_get_time` is ran after `super()._parse_parameter_file()`, the default value `-1` has already been set.

Reply:
> Although the value of current_time is printed by yt_logger after load, I thought it can tell users the current_time isn't set correctly due to lacking log file, not due to other bugs.

if you want to warn on stuff that users can't act upon, please use the yt logger itself. `warnings.warn` should be reserved to problems that users are able to fix on their end. Not having a log file doesn't seem like something users could fix.

Reply:
> Because _get_time is ran after super()._parse_parameter_file(), the default value -1 has already been set.

True, but it's not obvious from looking at the code here. It could also be a different value if something changes upstream, in which case your warning message would be wrong. You could mitigate this by inserting the value at runtime.

Reply:
Okay, now i get it.